tests/js/client/shell/shell-keygen.js

```diff
@@ -1383,6 +1369,287 @@ function KeyGenerationLocationSuite() {
+      IM.debugSetFailAt("KeyGenerator::generateOnSingleServer");
+
+      generators().forEach((generator) => {
+        let c = db._create(cn, {keyOptions: {type: generator}, numberOfShards: 2});
+        try {
+          // UPSERT should not fail because key generation happens with
+          // the coordinator key generator (potentially on the DBServer!)
+          let result = db._query(`UPSERT {value:1} INSERT {Hello:1, value:1} UPDATE {value:1} INTO ${cn} RETURN NEW`);
+          assertEqual(1, result.toArray().length);
+          assertEqual(1, c.count());
+        } finally {
+          db._drop(cn);
+        }
+      });
+    },
+
+    testUpsertSingleShardEdgeCollection: function () {
+      if (!cluster) {
+        return;
+      }
+
+      // fail if we generate a key with the single server key generator
+      IM.debugSetFailAt("KeyGenerator::generateOnSingleServer");
+
+      generators().forEach((generator) => {
+        let c = db._createEdgeCollection(cn, {keyOptions: {type: generator}, numberOfShards: 1});
+        try {
+          // UPSERT should not fail because key generation happens on coordinator
+          let result = db._query(`UPSERT {value:1} INSERT {Hello:1, value:1, _from: "${vn}/test", _to: "${vn}/test"} UPDATE {value:1} INTO ${cn} RETURN NEW`);
+          assertEqual(1, result.toArray().length);
+          assertEqual(1, c.count());
+        } finally {
+          db._drop(cn);
+        }
+      });
+    },
+
+    testUpsertMultiShardEdgeCollection: function () {
+      if (!cluster) {
+        return;
+      }
+
+      // fail if we generate a key with the single server key generator
+      IM.debugSetFailAt("KeyGenerator::generateOnSingleServer");
+
+      generators().forEach((generator) => {
+        let c = db._createEdgeCollection(cn, {keyOptions: {type: generator}, numberOfShards: 2});
+        try {
+          // UPSERT should not fail because key generation happens with the
+          // coordinator key generator
+          let result = db._query(`UPSERT {value:1} INSERT {Hello:1, value:1, _from: "${vn}/test", _to: "${vn}/test"} UPDATE {value:1} INTO ${cn} RETURN NEW`);
+          assertEqual(1, result.toArray().length);
+          assertEqual(1, c.count());
+        } finally {
+          db._drop(cn);
+        }
+      });
+    },
+
+    testUpsertNonStandardShardingCollection: function () {
+      if (!cluster) {
+        return;
+      }
+
+      // fail if we generate a key with the single server key generator
+      IM.debugSetFailAt("KeyGenerator::generateOnSingleServer");
+
+      generators().forEach((generator) => {
+        let c = db._create(cn, {keyOptions: {type: generator}, numberOfShards: 2, shardKeys: ["id"]});
+        try {
+          // UPSERT should not fail because key generation happens on coordinator
+          let result = db._query(`UPSERT {value:1} INSERT {Hello:1, value:1, id: "test"} UPDATE {value:1} INTO ${cn} RETURN NEW`);
```

Comment:
I would ask for a second upsert here, that includes the "id" also in the search pattern, this may trigger different query optimization.